gpcontrib/arenadata_toolkit/src/tf_shmem.c

```diff
@@ -0,0 +1,215 @@
+#include "bloom_set.h"
+#include "tf_shmem.h"
+
+#include "storage/ipc.h"
+#include "storage/shmem.h"
+
+#include "arenadata_toolkit_guc.h"
+
+#include <math.h>
+
+static shmem_startup_hook_type next_shmem_startup_hook = NULL;
+tf_shared_state_t *tf_shared_state;
+LWLock *tf_state_lock;
+LWLock *bloom_set_lock;
+tf_entry_lock_t bloom_locks[MAX_DB_TRACK_COUNT];
+uint64 bloom_hash_seed;
+int bloom_hash_num;
+
+/*
+ * Separate initialization of LWLocks;
+ */
+static void
+init_lwlocks(void)
+{
+	tf_state_lock = LWLockAssign();
+	bloom_set_lock = LWLockAssign();
+
+	for (int i = 0; i < db_track_count; ++i)
+	{
+		bloom_locks[i].lock = LWLockAssign();
+		bloom_locks[i].dbid = InvalidOid;
+	}
+}
+
+/*
+ * Calculate the closes power of 2 for given
+ * length.
+ */
+static int
+my_bloom_power(uint64 target_bitset_bits)
+{
+	int			bloom_power = -1;
+
+	while (target_bitset_bits > 0 && bloom_power < 32)
+	{
+		bloom_power++;
+		target_bitset_bits >>= 1;
+	}
+
+	return bloom_power;
+}
+
+/*
+ * Initialize optimal Bloom filter parameters
+ *
+ * This function calculates and sets optimal parameters for the Bloom filter
+ * based on established widespread principles.
+ *
+ * Adjusts the filter size to the nearest power of 2 to optimize memory alignment
```

Comment:
Current implementation aligns down the size. So, for ex., original bloom_size of 1023 will be transformed to 512 (twice less than requested). Is it Ok?

Reply:
Decided to work with size as it is